test/recipes/15-test_out_option.t

```diff
@@ -16,46 +16,51 @@ use OpenSSL::Test::Utils;
 
 setup("test_out_option");
 
-plan skip_all => "'-out' option tests are not available on Windows"
-    if $^O eq 'MSWin32';
-
-plan tests => 11;
-
-# The following patterns should be tested:
-#
-# path        dirname
-# /usr/       /
-# /           /
-# .           .
-# ..          .
-
-test_illegal_path('/usr/');
-test_illegal_path('/');
-test_illegal_path('./');
-test_illegal_path('../');
+# Paths that should generate failure when trying to write to them.
+# Directories are a safe bet for failure on all platforms.
+# Note that directories must end with a slash here, because of how
+# File::Spec massages them into directory specs on some platforms.
+my @failure_paths = (
+    './',
+   );
+my @success_paths = (
+    'test.pem'
+   );
 
 # Test for trying to create a file in a non-exist directory
-my @chars = ("A".."Z", "a".."z", "0".."9");
 my $rand_path = "";
-$rand_path .= $chars[rand @chars] for 1..32;
+do {
+    my @chars = ("A".."Z", "a".."z", "0".."9");
+    $rand_path .= $chars[rand @chars] for 1..32;
+} while (-d File::Spec->catdir('.', $rand_path));
 $rand_path .= "/test.pem";
 
-test_illegal_path($rand_path);
-test_legal_path('test.pem');
-unlink 'test.pem';
+push @failure_paths, $rand_path;
 
-sub test_illegal_path {
-    my $path = File::Spec->canonpath($_[0]);
-
-    my $start = time();
-    ok(!run(app([ 'openssl', 'genrsa', '-out', $path, '16384'])), "invalid output path: $path");
-    my $end = time();
-    # The above process should exit in 2 seconds if the path is not valid
-    ok($end - $start < 2, "check time consumed");
+# Specifically for mingw, the NULL device might not be what's expected.
+# For example, if cross compiled and tested on the build host, perl will
+# generate an incorrect NULL device name.
+# We might expand the exceptions...
+unless (config('target') =~ m|^mingw| && $^O ne 'msys') {
+    note "config('target') = ",config('target')," on line ",__LINE__;
+    note "\$^O = ",$^O," on line ",__LINE__;
+    # Check that we can write to the NULL device
+    push @success_paths, File::Spec->devnull();
 }
 
-sub test_legal_path {
-    my $path = File::Spec->canonpath($_[0]);
+plan tests => scalar @failure_paths + scalar @success_paths;
+
+foreach (@failure_paths) {
+    my $path = File::Spec->canonpath($_);
+    ok(!run(app([ 'openssl', 'genrsa', '-out', $path, '2048'])),
```

Comment:
Does it have be such expensive operation? Of course it's hardly noticeable on your desktop, but do you have to make Raspberry Pi user suffer [more than necessary]? rand -out $path would do... Or why not convert something from pem to der?

Reply:
Well, I suppose it applies more to success tests. That is if we assume that failure test would fail early....

Reply:
The reason for it is that this was already in the recipe before I started fiddling with it, and I simply didn't think further.  Of course we can change that to something better suited for quick checks...